### PR TITLE
cleanup Agent RSI handling, make loadRecording async

### PR DESCRIPTION
**Pull request by birarda**
- changed `Recording.loadRecording` to be asynchronous and take a callback
- fixed dependency teardowns in Agent
- fixed invokability of Agent.isAvatar calls

### CODE REVIEW

File: libraries/shared/src/SharedUtil.cpp [hunk 1]
```diff
@@ -769,7 +769,7 @@ bool similarStrings(const QString& stringA, const QString& stringB) {
 
 void disableQtBearerPoll() {
     // to work around the Qt constant wireless scanning, set the env for polling interval very high
-    const QByteArray EXTREME_BEARER_POLL_TIMEOUT = QString::number(INT_MAX).toLocal8Bit();
+    const QByteArray EXTREME_BEARER_POLL_TIMEOUT = QString::number(INT16_MAX).toLocal8Bit();
```
Comment (birarda) [inline libraries/shared/src/SharedUtil.cpp]: This change cleans up a debug message when running our targets over wifi, Qt must have changed their max allowable bearer timeout.

**Comment by birarda** [comment]
#### Test Plan

Use this AC script to test a looping recording.
https://gist.githubusercontent.com/birarda/0f5fcf77c1439b6a9d9cae385825c67d/raw/c01510ac855b182fe566cfe80ed87c86efacba23/ACPlaybackScript-Rec1.js

Add this script to your domain's persistent scripts. You should get an NPC playing a looping recording in your domain.

Change the persistent script in your domain settings to: https://gist.githubusercontent.com/birarda/bc927a637899410a373b1e53511b5350/raw/63cecbeadbef9c05c0843bfa4e963868e58285cc/ACPlaybackScript-Rec2.js.

Your domain should restart. When it returns the NPC should be playing a different looping recording.

You may need to use the PAL to put the orange orbs around speaking avatars, to find the recorded agents in your domain.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c7a28a527a2f487ad5534189e17690d3231586ba available at: https://gist.github.com/8958d31ffc6d90e743be97f901948b14<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10121-c7a28a5.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10121-c7a28a5.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1b245387565bd141ce467bf7f392d584ef8fe1ef available at: https://gist.github.com/8958d31ffc6d90e743be97f901948b14<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10121-1b24538.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10121-1b24538.dmg

**Comment by birarda** [comment]
When this is merged @RyanDowne we will need to change the bot scripts you are using.

File: script-archive/vrShop/review/shopReviewerAC.js [hunk 2]
```diff
@@ -46,21 +46,25 @@ function getAction(channel, message, senderID) {
...
-                
+
             case SHOW:
                 print("Show");
-                Recording.loadRecording(clip_url);
-                Agent.isAvatar = true;
-                Recording.setPlayerTime(0.0);
-                Recording.startPlaying();
-                Recording.stopPlaying();
+                Recording.loadRecording(clip_url, function(success){
```
Comment (thoys) [inline script-archive/vrShop/review/shopReviewerAC.js]: fix spacing here please

File: libraries/script-engine/src/RecordingScriptingInterface.h [hunk 3]
```diff
@@ -26,8 +28,11 @@ class RecordingScriptingInterface : public QObject, public Dependency {
 public:
     RecordingScriptingInterface();
 
+    void setScriptEngine(QScriptEngine* scriptEngine) { _scriptEngine = scriptEngine; }
+
 public slots:
-    bool loadRecording(const QString& url);
+
```
Comment (thoys) [inline libraries/script-engine/src/RecordingScriptingInterface.h]: Why the extra line?

File: libraries/script-engine/src/RecordingScriptingInterface.cpp [hunk 4]
```diff
@@ -52,31 +52,47 @@ float RecordingScriptingInterface::playerLength() const {
...
+                callback.call(_scriptEngine->globalObject(), args);
+            }
+
+            // drop our strong pointer to this clip so it is cleaned up
+            _clipLoaders.remove(clipLoader);
+        }
+    });
+
+    // when clip load fails, call the callback with the URL and failure boolean
+    connect(clipLoader.data(), &recording::NetworkClipLoader::failed, this, [this, weakClipLoader, url, callback](QNetworkReply::NetworkError error) mutable {
```
Comment (thoys) [inline libraries/script-engine/src/RecordingScriptingInterface.cpp]: max length of line should be max 128 chars

File: libraries/script-engine/src/RecordingScriptingInterface.cpp [hunk 5]
```diff
@@ -52,31 +52,47 @@ float RecordingScriptingInterface::playerLength() const {
...
-        loop.exec();
-    }
+    auto weakClipLoader = clipLoader.toWeakRef();
 
-    if (!loader->isLoaded()) {
-        qWarning() << "Clip failed to load from " << url;
-        return false;
-    }
+    // when clip loaded, call the callback with the URL and success boolean
+    connect(clipLoader.data(), &recording::NetworkClipLoader::clipLoaded, this, [this, weakClipLoader, url, callback]() mutable {
```
Comment (thoys) [inline libraries/script-engine/src/RecordingScriptingInterface.cpp]: max length of line should be max 128 chars

File: assignment-client/src/Agent.cpp [hunk 6]
```diff
@@ -753,8 +763,19 @@ void Agent::aboutToFinish() {
...
+    DependencyManager::destroy<ScriptCache>();
     DependencyManager::destroy<ScriptEngines>();
 
+    DependencyManager::destroy<ResourceCacheSharedItems>();
+    DependencyManager::destroy<SoundCache>();
+    DependencyManager::destroy<AudioScriptingInterface>();
+
+    DependencyManager::destroy<recording::Deck>();
+    DependencyManager::destroy<recording::Recorder>();
+    DependencyManager::destroy<recording::ClipCache>();
```
Comment (thoys) [inline assignment-client/src/Agent.cpp]: 👍 thanks for cleaning up

File: script-archive/tests/playbackAcTest.js [hunk 7]
```diff
@@ -2,17 +2,19 @@
...
 Avatar.skeletonModelURL = "https://hifi-public.s3.amazonaws.com/marketplace/contents/dd03b8e3-52fb-4ab3-9ac9-3b17e00cd85d/98baa90b3b66803c5d7bd4537fca6993.fst"; //lovejoy
 Avatar.displayName = "AC Avatar";
 Agent.isAvatar = true;
 
 Script.setTimeout(function () {
     Avatar.position = origin;
-    Recording.loadRecording("d:/hifi.rec");
-    Recording.setPlayerLoop(true);
-    Recording.startPlaying();
+    Recording.loadRecording("d:/hifi.rec", function(success){
```
Comment (thoys) [inline script-archive/tests/playbackAcTest.js]: ) { <-- spacing

File: script-archive/acScripts/PlayRecordingOnAC.js [hunk 8]
```diff
@@ -20,7 +20,11 @@ Avatar.orientation = Quat.fromPitchYawRollDegrees(0, 0, 0);
 Avatar.scale = 1.0;
 Agent.isAvatar = true;
 
-Recording.loadRecording(recordingFile);
+Recording.loadRecording(recordingFile, function(success) {
+    if (success) {
+        Script.update.connect(update);
+    }
```
Comment (thoys) [inline script-archive/acScripts/PlayRecordingOnAC.js]: I like the print message on fail solution in BetterClientSimulationBotFromRecording.js, could you do that here aswell?

File: script-archive/acScripts/triggeredRecordingOnAC.js [hunk 9]
```diff
@@ -42,10 +42,13 @@ var playRecording = function() {
...
-    Recording.loadRecording(CLIP_URL);
-    Recording.startPlaying();
-    isPlaying = true;
-    isPlayable = false; // Set this true again after the cooldown period
+    Recording.loadRecording(CLIP_URL, function(success) {
+        if (success) {
+            Recording.startPlaying();
+            isPlaying = true;
+            isPlayable = false; // Set this true again after the cooldown period
+        }
```
Comment (thoys) [inline script-archive/acScripts/triggeredRecordingOnAC.js]: I like the print message on fail solution in BetterClientSimulationBotFromRecording.js, could you do that here aswell?

File: script-archive/tests/playbackAcTest.js [hunk 10]
```diff
@@ -2,17 +2,19 @@
...
+        Recording.setPlayerLoop(true);
+        Recording.startPlaying();
```
Comment (thoys) [inline script-archive/tests/playbackAcTest.js]: What if the loading fails?

File: script-archive/vrShop/cash/shopCashierAC.js [hunk 11]
```diff
@@ -49,16 +49,19 @@ function update(deltaTime) {
...
-            Agent.isAvatar = true;
+            Recording.loadRecording(clip_url, function(success) {
+                if (success) {
+                    Recording.setPlayFromCurrentLocation(playFromCurrentLocation);
+                    Recording.setPlayerUseDisplayName(useDisplayName);
+                    Recording.setPlayerUseAttachments(useAttachments);
+                    Recording.setPlayerUseHeadModel(false);
+                    Recording.setPlayerUseSkeletonModel(useAvatarModel);
+                    Agent.isAvatar = true;
+                }
```
Comment (thoys) [inline script-archive/vrShop/cash/shopCashierAC.js]: I like the print message on fail solution in BetterClientSimulationBotFromRecording.js, could you do that here aswell? _(copy-pasted)_

File: scripts/developer/utilities/record/recorder.js [hunk 12]
```diff
@@ -281,8 +281,11 @@ function mousePressEvent(event) {
         if (!Recording.isRecording() && !Recording.isPlaying()) {
             recordingFile = Window.browse("Load recording from file", ".", "Recordings (*.hfr *.rec *.HFR *.REC)");
             if (!(recordingFile === "null" || recordingFile === null || recordingFile === "")) {
-                Recording.loadRecording(recordingFile);
-                setDefaultPlayerOptions();
+                Recording.loadRecording(recordingFile, function(success) {
+                    if (success) {
+                        setDefaultPlayerOptions();
+                    }
```
Comment (thoys) [inline scripts/developer/utilities/record/recorder.js]: I like the print message on fail solution in BetterClientSimulationBotFromRecording.js, could you do that here as well?

File: libraries/recording/src/recording/ClipCache.cpp [hunk 13]
```diff
@@ -21,18 +24,28 @@ void NetworkClip::init(const QByteArray& clipData) {
...
+        QMetaObject::invokeMethod(this, "getClipLoader", Qt::BlockingQueuedConnection,
+                                  Q_RETURN_ARG(NetworkClipLoaderPointer, result), Q_ARG(const QUrl&, url));
+        return result;
+    }
+
+    return getResource(url).staticCast<NetworkClipLoader>();
 }
 
 QSharedPointer<Resource> ClipCache::createResource(const QUrl& url, const QSharedPointer<Resource>& fallback, const void* extra) {
+    qDebug() << "Loading recording at" << url;
```
Comment (thoys) [inline libraries/recording/src/recording/ClipCache.cpp]: please put this debug message under category recordingLog, which is declared in Logging.h

**Comment by hifi-gustavo** [comment]
Build console output for commit: ec6f669c4283e7fb7a15dd04b9b9fe42eefc5397 available at: https://gist.github.com/8958d31ffc6d90e743be97f901948b14<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10121-ec6f669.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10121-ec6f669.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7b48feca662b6f43bf2a641e5ab731c203c4d8ad available at: https://gist.github.com/8958d31ffc6d90e743be97f901948b14<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10121-7b48fec.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10121-7b48fec.dmg

**Pull request merged**